PLUGINS.md

```diff
@@ -0,0 +1,447 @@
+    # that use alternative progress displays
+    # def register_status_indicator(function: Callable) -> None:
+    #     pass
+
+    # Register page for file properties. Same for track and album
+    # def register_file_info_page(page_class):
+    #     pass
+
+    # For the media player toolbar?
+    # def register_toolbar(toolbar_class):
+    #     pass
+```
+
+
+### Localization
+TBD
+
+
+### Plugin life cycle
+TBD
+
+
+## Distribution
+
+In order to both simplify third-party development of plugins and distribute
+the maintenance work there will no longer be a single plugin repository. Instead
+each plugin SHOULD be provided in a separate git repository. Plugins also CAN
+be installed locally without a git repository by placing the plugin package
+inside the plugin directory.
+
+
+### Repository structure
+
+A Picard plugin repository MUST be a git repository containing exactly one
+plugin. The content of the git repository MUST match the plugin file structure
+as described above, containing at least the `__init__.py` and `MANIFEST.toml`
+files.
+
+
+### Installation and upgrade
+
+Plugin installation is performed directly from git by cloning the git repository.
+Likewise updates are performed by updating the repository and checking out the
+requested git ref.
+
+For plugins installed from git the version will be shown as a combination of
+the version from the manifest and the git ref (`{VERSION}-{GITREF}`).
+
+
+### Official plugins
+
+The Picard website will provide a list of officially supported plugins and their
+git location. Those plugins will be offered in the Picard user interface for
+installation. Plugins can be added to the official list after a review. The
+Picard website must provide an API endpoint for querying the metadata for all
+the plugins. The metadata consists of both the information from the plugin
+manifests and the git URL for each plugin.
+
+The exact implementation for submitting plugins for the Picard website is
+outside the scope of this document and will be discussed separately. It could
+e.g. both be handled by opening tickets on the MetaBrainz Jira or by
+implementing an actual plugin submission interface directly on the Picard
+website.
+
+
+### Installing plugins from unofficial sources
+
+Picard must provide a user interface for installing third-party plugins which
+are not provided in the official plugin list. The user needs to enter the
+plugin's git URL and Picard will verify the manifest and offer to install and
+activate the plugin. The UI must make it clear that the user is installing the
+plugin at their own risk and that the plugin can execute arbitrary code.
+
+
+## To be discussed
+
+### Localization
+
+Existing plugins in Picard 2 cannot be localized. The new plugin system should
+allow plugins to provide translations for user facing strings.
+
+Plugins could provide gettext `.mo` files that will be loaded under a plugin
+specific translation domain.
```

Comment:
How would that work?

1. Plugins have their own domain, but cannot access picard domains, this may be a problem for constants/attributes from database
2. Plugins have their own domain, a set of methods from plugin API to use it (_(), N_() are those from Picard, P_() and PN_() for plugins)
3. We append plugin domain to the main one (chaining fallbacks https://docs.python.org/3/library/gettext.html#gettext.NullTranslations.add_fallback), basically Picard main domain is used, but plugin "extends" it, standard picard/i18n are used
4. ?

We need to test each approach to find the best.
Loading plugin translations can be either automatic (we expect .mo files to be somewhere identical for all plugins), or by a specific call from the plugin to the API to indicate where to find translations  (decided by plugin author). I tend to prefer the first approach, as it is simple, safer, and "normalize" plugins.

Reply:
My preference would also be 1. I think plugins should have their own translation generally. For thing like attribute translations we have separate domains anyway. If this becomes a requirement we could consider exposing this via the API as well.

Opiton 3. sounds interesting though, we should probably try how well this works.

Reply:
Actually using fallbacks isn't a good idea, as it may leads to conflicts between strings (if we search main domain first and plugin has a different translation for the same string, it will not appear, and the reverse searching plugin domain first and falling back on main, would let plugin translations to override main translations).

Also `.mo` aren't supposed to be portable between systems.
Though GNU gettext is able to take care of endianess at runtime it seems, it is usually preferable to compile .po to .mo on target system.
An interesting thread about that: https://stackoverflow.com/questions/53285634/is-there-a-portable-way-to-provide-localization-of-a-package-distributed-on-pypi